plugins/module_utils/role_utils.py

```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2019, XLAB Steampunk <steampunk@xlab.si>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+def _sort_rules(rules):
+    for rule in rules:
+        for key in rule:
+            if rule[key]:
+                rule[key].sort()
+    return rules
+
+
+def _do_rules_differ(current_rules, desired_rules):
+    sorted_current = _sort_rules(current_rules)
+    sorted_desired = _sort_rules(desired_rules)
+
+    if len(sorted_current) != len(sorted_desired):
+        return True
+
+    for rule in sorted_desired:
+        if rule not in sorted_current:
+            return True
+
+    return False
+
+
+def do_differ(current, desired):
+    if current is None:
+        return True
+
+    for key, value in desired.items():
+        if key == 'rules':
+            if _do_rules_differ(current['rules'], desired['rules']):
+                return True
+        elif value != current.get(key):
```

Comment:
Although I really like how nice and readable L38 is, we already have `desired['rules']` stored in `value` and we could also potentially store `current['rules']` (which is exactly `current.get(key)` from L40) to `current_value`. 

Something like:
```python
for key, value in desired.items():
    current_value = current.get(key)
    if key == 'rules':
        if _do_rules_differ(current_value, value):
            return True
    elif value != current_value:
           return True
```
I don't mind if you leave it as is though, I will have to slightly modify this part myself in #83 :slightly_smiling_face:.

Reply:
Heh, that condition was not in the loop at first, that's why it looks like this :sweat_smile:. I moved it there few seconds before the push. I'll change it, it looks better anyway :+1: